test/thx/schema/TestGeneric.hx

```diff
@@ -0,0 +1,416 @@
+  }
+
+  public function testMaybeField() {
+    var s = schema(WithMaybe)(float());
+    roundTripSchema({ maybe: thx.Maybe.none() }, s);
+    roundTripSchema({ maybe: thx.Maybe.of(0.1) }, s);
+  }
+
+  // TODO
+  // public function testGadt() {
+  //   var s = schema(GADT)();
+  //   roundTripSchema(A(1), s);
+  //   roundTripSchema(B("b"), s);
+  // }
+}
+
+typedef WithOption<T> = {
+  opt: Option<T>
+}
+
+typedef WithNullable<T> = {
+  nullable: Null<T>
+}
+
+typedef WithMaybe<T> = {
+  maybe: thx.Maybe<T>
+}
+
+typedef TypeWithMapDef = { a: Map<String, Int> };
+typedef TypeWithMap<T> = { a: Map<String, T> };
+typedef TypeWithMapInt = { a: Map<String, Int> };
+typedef TypeWithMaybe = { a: thx.Maybe<Int> };
+
+class SimpleClass {
+  var a: String;
+  var b: Int;
+  public function new(a: String, b: Int) {
+    this.a = a;
+    this.b = b;
+  }
+}
+
+class InnerClass {
+  var a: String;
+  public function new(a: String) {
+    this.a = a;
+  }
+}
+
+class OuterClass {
+  var b: String;
+  var inner: InnerClass;
+  public function new(b: String, inner: InnerClass) {
+    this.b = b;
+    this.inner = inner;
+  }
+}
+
+class ClassWithTypeParameters<A, B> {
+  var a: A;
+  var b: B;
+  var c: Int;
+  public function new(a: A, b: B, c: Int) {
+    this.a = a;
+    this.b = b;
+    this.c = c;
+  }
+}
+
+class RecursiveClass<X> {
+  var r: Option<RecursiveClass<X>>;
+  var x: X;
+  public function new(r: Option<RecursiveClass<X>>, x: X) {
+    this.r = r;
+    this.x = x;
+  }
+}
+
+class EmptyClass {
+  public function new() {}
+}
+
+class Class1 {
+  public var s: String;
+  var s1: String = "z";
+  public var s2(null, null) = "z";
+  public var i = 3;
+  public var f(default, null) = 3.0;
+  public var f1(get, null): Float;
+  @:isVar public var f2(get, set): Float;
+
+  public function new(s) {
+    this.s = s;
+    f1 = 3;
+    f2 = 4.0;
+  }
+
+  public function getI() {
+    return this.i;
+  }
+
+  function get_f1() {
+    return f1;
+  }
+
+  function get_f2() {
+    return f2;
+  }
+
+  function set_f2(v: Float) {
+    return f2 = v;
+  }
+}
+
+class Class2 {
+  public var myInt: MyInt;
+  public var myOpt: Option<Either<String, Array<Float>>>;
+
+  public function new(myInt, myOpt) {
+    this.myInt = myInt;
+    this.myOpt = myOpt;
+  }
+}
+
+class Class3<T1, T2> {
+  public var t1: T1;
+  public var t2: T2;
+
+  public function new(t1, t2) {
+    this.t1 = t1;
+    this.t2 = t2;
+  }
+}
+
+enum Case1<T1, T2> {
+  A;
+  B(bs: String);
+  C(cs: String, ci: Int, cf: Float, cb: Bool);
+  D(d: MyInt);
+  E(e: Array<String>);
+  F(f: Array<Float>);
+  G(a: T1);
+  H(a: Array<Array<MyInt>>);
+  I(a: Array<T1>);
+  J(e: Either<Int, Float>);
+  K(t1: T1, t2: T2);
+  L(t1: Option<String>);
+  M(s: Null<String>);
+  N(?s: String);
+  O(e: Either<T1, Float>);
+  P(v: Option<Case1<T1, T2>>);
+  // O<T3>(t3: Option<Array<T3>>);
+}
+
+enum Case2 {
+  A;
+  B(di: MyInt, z: haxe.ds.Option<String>);
+}
+
+enum GADT<T> {
+  A(a: T): GADT<Int>;
```

Comment:
Just a general question here - does `T` unify with `Int` in this case (and `String`) below? I haven't seen a GADT type variable in this position before.